lib/src/steps/step.dart

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:meta/meta.dart' show sealed;
+import 'package:file/file.dart';
+import 'result.dart' show ResultBuilder, BreakStep;
+
+export 'result.dart' show ResultBuilder, BreakStep;
+
+@sealed
+class Step {
+  /// Unique name of this analysis step.
+  ///
+  /// This **must** be a unique lowercase identifier matching `^[a-z0-9_-]+$`.
+  final String name;
+
+  /// Run analysis given an [AnalysisContext].
+  ///
+  /// This may throw [BreakStep] to report a final suggestion and stop analysis.
```

Comment:
Wondering about this: why should we combine the final suggestion with the exception?

Reply:
Throwing it is optional...

It allows us to make helper functions that throws an exception that stops analysis for the given step and sets a suggestion, without access to the context...

It's heavily inspired by what we did with `ResponseException` in pub-dev.

I'm realizing now, we should stop prefixing `kind` with the step name. And then we should remove duplicate suggestions...

Then we can make a function that reads and parses the `pubspec.yaml`. If it fails it throw a `BreakStep` exception, thus, a step that uses this function doesn't need to check if the return value is `null`, nor does it need to catch the exception. Most steps that need a parsed pubspec will just give up doing anything else and return.

Reply:
Note to self: refactor Step into having a constructor that creates a single step from a list of steps...

Then I can also simplify to `runStep` instead of `runSteps`.

Reply:
> It allows us to make helper functions that throws an exception that stops analysis for the given step and sets a suggestion, without access to the context...

Please add this paragraph to the comment block.